### DISCUSSION TITLE
how to hide mo file code？

### DISCUSSION THREAD

**Discussion by pily1** [OpenModelica/OpenModelica#12479] [Q&A] [answered]
hi，I want to hide some function code,the code is writed by modelica language,how can I hide the code?

**Comment by adeas31** [comment]
You need to add protection and access control annotation to your code and then use encryption on top of it. Read https://specification.modelica.org/master/annotations.html#access-control-to-protect-intellectual-property on how to use the annotations. OpenModelica support encrypted libraries using https://github.com/modelica/Encryption-and-Licensing

**Reply by pily1** [reply]
Can the mo file be compiled into a dynamic library ? if can,how to do it?

**Reply by arun3688** [reply]
@pily1 Yes it is possible to compile into a dll, you can export the model into an fmu which generates the dll or you can use openmodelica scripting to generate the dll for example
 
Test.mo 
```
model Test
end Test;
```
a.mos
 ```
loadFile("Test.mo");
getErrorString();
translateModel(Test); 
getErrorString();
system("mingw32-make -f  Test.makefile omc_dll_target -j5, Test_compile.log") // this will generate the dll of your model 
getErrorString():

```

**Reply by pily1** [reply]
hi,I have tried it,but get error,the following image is the error,how to solve it?
E:\work\openModelica\test>D:\OpenModelica\bin\omc.exe a.mos
Error processing file: a.mos
[E:/work/openModelica/test/a.mos:6:1-6:1:writable] Error: Missing token: ASSIGN

# Error encountered! Exiting...
# Please check the error message and the flags.

Execution failed!

**Reply by arun3688** [reply]
@pily1 semicolons are missing in the mos script,use the below

```
loadFile("Test.mo");
getErrorString();
translateModel(Test); 
getErrorString();
system("mingw32-make -f  Test.makefile omc_dll_target -j5, Test_compile.log"); // this will generate the dll of your model 
getErrorString();
```

**Reply by pily1** [reply]
@arun3688 this one report no error,but it generate a lot of c language files without dynamic library,the following image is the output files;
![1111](https://github.com/OpenModelica/OpenModelica/assets/20216269/9980207e-f4fb-4454-b047-55c9114c7517)

**Reply by arun3688** [reply, marked as answer]
Run the following script and it should work and make sure you have the latest OpenModelica version probably nightly builds, because this feature does not exist in OpenModelica 1.22 
```
loadFile("Test.mo");
getErrorString();
translateModel(Test);
getErrorString();
system("mingw32-make -f Test.makefile omc_dll_target -j5"); // this will generate the dll of your model
getErrorString();
```

**Reply by pily1** [reply]
@arun3688 think you,it  has generated dynamic library correctly.another question,how to use the dll's model?is the same with external C language library?C language has head file,but mo file has not head file.

**Comment by arun3688** [comment]
@pily1 yes it is same like external C language library, the entry point to the dll is  `omcDllMain`, So you have to load the dll and call the function something like this          

`   int result = omcDllMain(myargs.Length, myargs);
`
I don't know what you want to do, if your aim is to just hide the functions in the model, then you can simple export the model as fmu with protected settings and it will generate the fmu with dll and you can use OMSimulator to simulate the fmu and get the results, I think that would be the easiest.
